src/compiler/checker.ts

```diff
@@ -12419,11 +12414,14 @@ namespace ts {
             }
             if ((<Identifier>node.name).text === "this") {
                 if (indexOf(func.parameters, node) !== 0) {
-                    error(node, Diagnostics.this_parameter_must_be_the_first_parameter);
+                    error(node, Diagnostics.A_this_parameter_must_be_the_first_parameter);
                 }
-                if (func.kind === SyntaxKind.Constructor) {
+                if (func.kind === SyntaxKind.Constructor || func.kind === SyntaxKind.ConstructSignature) {
                     error(node, Diagnostics.A_constructor_cannot_have_a_this_parameter);
                 }
+                if (func.kind === SyntaxKind.SetAccessor) {

```

Comment:
you have construct signature covered, i.e. `{ new(this: T): T; }`, i am talking about `ConstructorType` i.e. `var x: new (this: number)=> number`. this still fails on the latest from this branch.